### DISCUSSION TITLE
How do I pass objetcs to "seqcli log -m" ?

### DISCUSSION THREAD

**Discussion by yooakim** [datalust/seq-tickets#1486] [Help]
I'm using quite a bit of PowerShell and it can sometimes be desirable to quickly log what you do in PowerShell.

One though I have is to use `seqcli` to simply log messages. Logging simple messages works fine, as this:

     seqcli log -m 'Bonjour' -p Environment=Production

But what I really would like is to be able to use a message template to capture structured messages. As an example:

```powershell
$processes = Get-Process
seqcli log -m 'Processes {@Processes}'
```

as I understand it the `secqli log` command [supports message templates](https://docs.datalust.co/docs/command-line-client#log). but what I can't figure out is how do I send the values to be used by the message template from `seqcli`? 

This may not be a good way do log from PowerShell, if so let me know! :-) but if one could send structured log messages from PowerShell (seqcli) it would be great.

**Comment by liammclennan** [comment]
Hi @yooakim,
Use `-p` to specify each property. E.g.

```
seqcli log -m 'Logging in {Environment}' -p Environment=Production
```

**Reply by yooakim** [reply]
Thanks @liammclennan , I did try that but that did not work if I pass in a PowerShell variable. What I was hoping is that it would work if I did like this:

    $processes = Get-Process
    seqcli log -m 'Processes {@Processes}' -p Processes=$processes 

but then the `seqcli` does not understand/expand the variable. It looks as this:
<img width="711" alt="image" src="https://user-images.githubusercontent.com/862546/154209652-87c480ea-e606-4464-95c8-753916b62342.png">

what I was hoping for is that the properties of each object in the $process list would be expanded to properties. 

The other way would be to simpley extract the values and add them as parameters, that would work but is more work on the PowwerShell side.  So this works

    $processes | ForEach-Object { seqcli log -m 'Processes {@Name} {@Path}' -p Name=$($_.ProcessName) -p Path=$($_.Path)}
    
 and it works fine:  
<img width="613" alt="image" src="https://user-images.githubusercontent.com/862546/154210482-430a9d95-0f8d-4452-b1b3-171b0764e1e8.png">

but ideally I could pass the object as in the first example :-)

**Reply by nblumhardt** [reply]
Hi @yooakim - unfortunately, because `seqcli` is an executable and not a PowerShell cmdlet, all of the expansion happens before `seqcli` sees anything, so it's not possible to make this case work, currently.

I don't remember having tried it 😁 but - you might actually be able capture structured property values in PowerShell using:

https://docs.datalust.co/docs/using-powershell

It's not based on Serilog, so you don't need `@`, but something like this would work:

```powershell
Send-SeqEvent $seq "Processes attached" -properties @{ Processes = $processes } -template
```

Except that processes are a large/cyclic structure, so converting one to JSON is likely to hang/burn CPU cycles. If you can create some temporary/anonymous objects with just the name/executable/PID etc. as desired, this might work.

HTH!

**Reply by yooakim** [reply]
Yes, thanks. I'll explore this but for now I think it's enought to simply add the parameters. Thanks. @nblumhardt  (yes, the process class/objects aren't a good example, I agree :-) )
